arc.go

```diff
@@ -0,0 +1,230 @@
+			var res [][]string
+			for _, arcSet := range arcSets {
+				res = append(res,
+					[]string{aarKey, arcSet.authenticationResults.Raw},
+					[]string{amsKey, arcSet.messageSignature.Raw},
+				)
+
+				if arcSet.messageSignature.ArcInstance == i+1 {
+					break
+				}
+				// skip last seal as it's not in the signature
+				res = append(res, []string{asKey, arcSet.seal.Raw})
+			}
+			return res
+		}
+
+		arcSets[i].seal.getHeadersFunc = getArcHeaders
+
+		res, err := arcSets[i].verify(i+1, msg)
+		if err != nil {
+			return &Result{Result: Fail, Error: &VerificationError{Source: VerifyError, Err: err}}, nil
+		}
+		results = append(results, *res)
+	}
+
+	arcResult := func(result ResultCode, msg string, i int) *Result {
+		return &Result{Result: Fail, Error: &VerificationError{
+			Source:      VerifyError,
+			Explanation: msg,
+			Tag:         "i",
+			Value:       strconv.Itoa(i),
+		}}
+	}
+
+	if !results[0].amsValid {
+		return arcResult(Fail, "Most recent ARC-Message-Signature did not validate", results[0].instance), nil
+	}
+
+	// Validate results
+	//
+	//	"The "cv" value for all ARC-Seal header fields MUST NOT be
+	//	"fail".  For ARC Sets with instance values > 1, the values
+	//	MUST be "pass".  For the ARC Set with instance value = 1, the
+	//	value MUST be "none"."
+	for _, res := range results {
+		switch {
+		case res.cv == Fail:
+			return arcResult(Fail, "ARC-Seal reported failure, the chain is terminated", res.instance), nil
+		case !res.asValid:
+			return arcResult(Fail, "ARC-Seal did not validate", res.instance), nil
+		case (res.instance == 1) && (res.cv != None):
+			return arcResult(Fail, "ARC-Seal reported invalid status", res.instance), nil
+		case (res.instance > 1) && (res.cv != Pass):
+			return arcResult(Fail, "ARC-Seal reported invalid status", res.instance), nil
+		}
+	}
+
+	return &Result{Result: Pass}, nil
+}
+
+func extractArcSets(headers MIMEHeader) ([]*arcSet, error) {
+	arcSeals := headers[canonicalMIMEHeaderKey([]byte(asKey))]
+	signatures := headers[canonicalMIMEHeaderKey([]byte(amsKey))]
+	results := headers[canonicalMIMEHeaderKey([]byte(aarKey))]
+
+	// Each arc-set must have exactly one of each header (seal, message signature and authentication results)
+	instances := len(arcSeals)
+	if (instances != len(signatures)) || (len(signatures) != len(results)) {
+		return nil, errMissingArcFields
+	}
+
+	sets := make([]*arcSet, instances)
+	for i := 0; i < instances; i++ {
+		as, err := parseSignature(asKey, arcSeals[i].Folded, arcSeals[i].Original, requiredASTags)
+		if err != nil {
+			return nil, err
+		}
+		// Seals only use "relaxed" header field canonicalization
+		// https://www.rfc-editor.org/rfc/rfc8617.html#section-4.1.3
+		as.RelaxedHeader = true
+
+		ams, err := parseSignature(amsKey, signatures[i].Folded, signatures[i].Original, requiredAMSTags)
+		if err != nil {
+			return nil, err
+		}
+		// if "c=" not given, use relaxed
+		if !ams.canonicalization {
+			ams.RelaxedHeader = true
```

Comment:
Won't this always be `true` already?

Reply:
On `as`, yes.  The `ams` is the result of `parseSignature(...)`

Reply:
Yea on AS, on the AMS header sometimes the cannonization field is not provided, so we just want a default here.